apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java

```diff
@@ -179,6 +182,23 @@ public Transaction startNoop() {
         return this;
     }
 
+    @Override
+    public Transaction activate() {
+        if (!wasActivated.get()) {
+            if (!wasActivated.getAndSet(true)) {
+                onFirstActivation();
+            }
+        }
```

Comment:
[minor] the outer `if` seems redundant here

Reply:
This is a small performance optimization: `get()` on an atomic is a volatile read, which costs the same as a normal read. `getAndSet()` is an atomic operation (similar to CAS) and therefore comes with the memory-barrier penalty like other atomic operations. I'll add a comment to clarify